libphx/src/engine.rs

```diff
@@ -26,157 +28,176 @@ pub static subsystems: u32 = SDL_INIT_EVENTS
+            SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_DOUBLEBUFFER, 1);
+            SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_DEPTH_SIZE, 24);
+
+            Keyboard_Init();
+            Metric_Reset();
+            Mouse_Init();
+            Input_Init();
+            Resource_Init();
+            ShaderVar_Init();
+
+            initTime = TimeStamp_Get();
         }
+    }
 
-        if SDL_Init(0) != 0 {
-            CFatal!("Engine_Init: Failed to initialize SDL");
-        }
-        if !Directory_Create(c_str!("log")) {
-            CFatal!("Engine_Init: Failed to create log directory.");
+    pub fn free() {
+        unsafe {
+            ShaderVar_Free();
+            Keyboard_Free();
+            Mouse_Free();
+            Input_Free();
+            Signal_Free();
+            SDL_QuitSubSystem(subsystems);
         }
-        atexit(Some(SDL_Quit as unsafe extern "C" fn() -> ()));
     }
 
-    if SDL_InitSubSystem(subsystems) != 0 {
-        CFatal!("Engine_Init: Failed to initialize SDL's subsystems");
+    pub fn abort() {
+        unsafe {
+            abort();
+        }
     }
 
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_CONTEXT_MAJOR_VERSION, glVersionMajor);
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_CONTEXT_MINOR_VERSION, glVersionMinor);
-    SDL_GL_SetAttribute(
-        SDL_GLattr::SDL_GL_CONTEXT_PROFILE_MASK,
-        SDL_GLprofile::SDL_GL_CONTEXT_PROFILE_COMPATIBILITY as i32,
-    );
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_ACCELERATED_VISUAL, 1);
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_RED_SIZE, 8);
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_GREEN_SIZE, 8);
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_BLUE_SIZE, 8);
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_DOUBLEBUFFER, 1);
-    SDL_GL_SetAttribute(SDL_GLattr::SDL_GL_DEPTH_SIZE, 24);
-
-    Keyboard_Init();
-    Metric_Reset();
-    Mouse_Init();
-    Input_Init();
-    Resource_Init();
-    ShaderVar_Init();
-
-    initTime = TimeStamp_Get();
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Engine_Free() {
-    ShaderVar_Free();
-    Keyboard_Free();
-    Mouse_Free();
-    Input_Free();
-    Signal_Free();
-    SDL_QuitSubSystem(subsystems);
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Engine_Abort() {
-    abort();
-}
-
-#[no_mangle]
-pub extern "C" fn Engine_GetBits() -> i32 {
-    8_usize.wrapping_mul(std::mem::size_of::<*mut libc::c_void>()) as i32
-}
+    pub fn get_bits() -> i32 {
+        8_usize.wrapping_mul(std::mem::size_of::<*mut libc::c_void>()) as i32
+    }
 
-#[no_mangle]
-pub unsafe extern "C" fn Engine_GetTime() -> f64 {
-    TimeStamp_GetElapsed(initTime)
-}
+    pub fn get_time() -> f64 {
+        unsafe { TimeStamp_GetElapsed(initTime) }
+    }
 
-#[no_mangle]
-pub extern "C" fn Engine_GetVersion() -> *const libc::c_char {
-    env!("PHX_VERSION").as_ptr() as *const libc::c_char
-}
+    pub fn get_version() -> &'static str {
+        // TODO: think about string conversion to C string without pinning if necessary
```

Comment:
I'm thinking, if we are returning a `&str`, then in theory we don't need to bother with the string pinning, however the problem is we need to convert that `&str` into a `CString` because `&str` is not null terminated, and we need to store that allocated `CString` somewhere. So I think the pinning is still required in this case.

Looks like the previous code `env!("PHX_VERSION").as_ptr() as *const libc::c_char` was broken as a result... (because that'snot necessarily a null terminated ptr).

I wonder whether it's possible for us to return a `&'static CStr` instead in some cases (although probably not here), and have that be turned into a ptr without pinning?

Reply:
I'll think about CStr and will add support for CString

Reply:
Added CString support